module/state_synchronization/requester/status/status.go

```diff
@@ -0,0 +1,281 @@
+	// LastNotified is the highest consecutive execution data received. When loaded, this is used
+	// as the starting point for requesting ExecutionData for new blocks.
+	// Note: LastNotified is used instead of LastReceived since there could be a large gap of
+	// un-downloaded blocks between LastReceived and LastNotified. Using LastNotified allows the
+	// requester to backfill all of missing blocks even when it's not configured to recheck the
+	// datastore.
+	LastNotified uint64
+
+	// Halted is persisted so the condition can be detected without inspecting the entire datastore
+	Halted bool
+}
+
+// New returns a new Status struct
+func New(db datastore.Batching, log zerolog.Logger, startHeight uint64, maxCachedEntries, maxSearchAhead uint64) *Status {
+	return &Status{
+		db:   db,
+		log:  log,
+		heap: &notificationHeap{},
+
+		startHeight:   startHeight,
+		lastNotified:  startHeight,
+		lastReceived:  startHeight,
+		lastProcessed: startHeight,
+
+		maxCachedEntries: maxCachedEntries,
+		maxSearchAhead:   maxSearchAhead,
+	}
+}
+
+// Load loads status from the db
+func (s *Status) Load(ctx context.Context) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	data, err := s.db.Get(ctx, datastore.NewKey(statusDBKey))
+	if err == datastore.ErrNotFound {
+		return nil
+	}
+
+	if err != nil {
+		return fmt.Errorf("failed to load status: %w", err)
+	}
+
+	saved := persistedStatus{}
+
+	err = json.Unmarshal(data, &saved)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal status: %w", err)
+	}
+
+	s.lastNotified = saved.LastNotified
+	s.halted = saved.Halted
+
+	// processing restarts after a fresh boot.
+	s.lastReceived = s.lastNotified
+	s.lastProcessed = s.lastNotified
+
+	return nil
+}
+
+// save writes status to the db
+func (s *Status) save(ctx context.Context) error {
+	savedStatus := persistedStatus{
+		LastNotified: s.lastNotified,
+		Halted:       s.halted,
+	}
+
+	data, err := json.Marshal(&savedStatus)
+	if err != nil {
+		return fmt.Errorf("failed to marshal status: %w", err)
+	}
+
+	err = s.db.Put(ctx, datastore.NewKey(statusDBKey), data)
+	if err != nil {
+		return fmt.Errorf("failed to save status: %w", err)
+	}
+
+	return nil
+}
+
+// NextNotification returns the BlockEntry for the next notification to send and marks it as notified.
+// If the next notification is available, it returns the BlockEntry and true. Otherwise, it returns
+// nil and false.
+func (s *Status) NextNotification(ctx context.Context) (*BlockEntry, bool) {
+	// This needs a write lock since it may modify the heap
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	var entry *BlockEntry
+	next := s.nextHeight()
+
+	// Remove any entries for the next height or below. This ensures duplicate or unexpected heights
+	// below the tracked heights are removed. If there are duplicate entries for the same height, the
+	// last entry popped from the heap will be returned.
+	for s.heap.Len() > 0 && s.heap.PeekMin().Height <= next {
+		entry = heap.Pop(s.heap).(*BlockEntry)
```

Comment:
Why not calling `s.heap.Pop()`?

Seems like leaking implementation detail

Reply:
this is the interface for the go heap implementation. under the covers, `heap.Pop` calls `s.heap.Pop()` and a method to rebalance.